### PR TITLE
Update amigo bot for puzzle 2

### PR DESCRIPTION
**Pull request by worldwise001**
r: @alokmenghrajani @csstaub

### CODE REVIEW

File: amigo_bot.go [hunk 1]
```diff
@@ -13,12 +13,14 @@ package main
...
-	"github.com/nlopes/slack"
-	"golang.org/x/net/websocket"
 	"log"
+	"strconv"
 	"strings"
 	"sync"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/nlopes/slack"
+	"golang.org/x/net/websocket"
```
Comment (csstaub) [inline amigo_bot.go:23]: yes to goimports

File: amigo_bot.go [hunk 2]
```diff
@@ -298,14 +349,19 @@ func doValidate(config Config, db *sql.DB, ws *websocket.Conn, userToken string,
...
 func doTopScores(config Config, db *sql.DB, ws *websocket.Conn, userToken string, channel string) {
-	rows, err := db.Query("select id, name, unix_timestamp(flag1)-unix_timestamp(start) as flag1_time, unix_timestamp(flag2)-unix_timestamp(start) as flag2_time from teams_start_flag1_flag2 where id < 666 and flag1 is not null order by -flag2_time desc, flag1_time")
+	// I'm sorry
+	rows, err := db.Query("select distinct teams.id as id, teams.name as team_name, unix_timestamp(t1.ts)-unix_timestamp(t0.ts) as flag1_time, unix_timestamp(t2.ts)-unix_timestamp(t0.ts) as flag2_time, FLOOR(unix_timestamp(t3.ts)/unix_timestamp(t3.ts)) AS flag3_time, (select count(*) from logs as a where a.team_id=teams.id and level=2 and (event like 'incorrect:%' or event like 'flag 3')) AS flag3_tries from logs left join teams on (teams.id=logs.team_id) left join logs as t0 on (t0.team_id=logs.team_id and t0.event='start') left join logs as t1 on (t1.team_id=logs.team_id and t1.event='flag 1') left join logs as t2 on (t2.team_id=logs.team_id and t2.event='flag 2') left join logs as t3 on (t3.team_id=logs.team_id and t3.event='flag 3') order by (flag3_time * -flag3_tries) desc, -flag2_time desc, flag1_time")
```
Comment (csstaub) [inline amigo_bot.go:363]: Does this sort NULLs for flag3 at the bottom? I initially had a bug where it sorted NULLs for flag2 at the top which was weird

File: amigo_bot.go [hunk 3]
```diff
@@ -267,14 +282,50 @@ func doValidate(config Config, db *sql.DB, ws *websocket.Conn, userToken string,
...
+	level, err = strconv.Atoi(sLevel)
+	switch {
+	case err != nil:
+		postError(ws, channel, fmt.Sprintf("%s is not a valid puzzle number", sLevel), userToken)
+		return
+	case level < 1:
+		postError(ws, channel, fmt.Sprintf("you give us too much credit for starting puzzle enumeration from 0; humans designed this, not chat bots"), userToken)
+		return
+	case level > 2:
+		postError(ws, channel, fmt.Sprintf("woaaaaah nelly! puzzle 3 hasn't started yet!"), userToken)
```
Comment (csstaub) [inline amigo_bot.go:294]: lol

File: amigo_bot.go [hunk 4]
```diff
@@ -298,14 +349,19 @@ func doValidate(config Config, db *sql.DB, ws *websocket.Conn, userToken string,
 		m.Text = fmt.Sprintf("Congrats, you found %s!", event)
 	} else {
 		m.Text = fmt.Sprintf("Sorry, that's not right.")
+		if level == 2 {
+			m.Text += fmt.Sprintf(" You have %d tries left.", 10-(count+1))
+		}
```
Comment (csstaub) [inline amigo_bot.go:354]: where do you update the counter?

File: amigo_bot.go (hunk 2 above)
Comment (worldwise001) [inline amigo_bot.go:363]: It turns out flag3_tries always is non-null, and flag3 is sometimes null, and if you multiply null with a value you get null, and the nulls somehow get sorted to the bottom.

Comment (csstaub) [inline amigo_bot.go:363]: seems good to me

File: amigo_bot.go (hunk 4 above)
Comment (worldwise001) [inline amigo_bot.go:354]: The counter is calculated with a SELECT COUNT(*), i.e. an insert with an event of incorrect will cause this to be bumped up automatically.

**Comment by alokmenghrajani** [comment]
👍

File: amigo_bot.go [hunk 5]
```diff
@@ -267,14 +282,50 @@ func doValidate(config Config, db *sql.DB, ws *websocket.Conn, userToken string,
...
+			event_ok = true
+		}
+	case level == 2:
+		// Make sure they haven't done > 10 tries
+		switch {
+		case count >= 10:
+			postError(ws, channel, fmt.Sprintf("you've exhausted your 10 tries! no points 4 u"), userToken)
+			return
+		default:
+			if flag == config.Flag3 {
```
Comment (alokmenghrajani) [inline amigo_bot.go:324]: nit: don't forget to include the changes to config.go and config.json.sample.

**Pull request merged**